File: name-inverter/name-inverter-py/06/NameInverter.py
import re
from typing import List


class NameInverter:
    @classmethod
    def invert(cls, name: str) -> str:
        name_parts = cls.break_into_parts_ignoring_whitespaces(name)
        if len(name_parts) < 2:
            return name_parts[0]
        return cls.invert_parts(cls.without_honorifics(name_parts))

    @staticmethod
    def break_into_parts_ignoring_whitespaces(name: str) -> List[str]:
        return re.split(RegularExpressions.any_amount_of_whitespaces(), name.strip())

    @staticmethod
    def without_honorifics(name_parts: List[str]) -> List[str]:
        if Honorifics.is_honorific(name_parts[0]):
            name_parts.pop(0)
        return name_parts

    @classmethod
    def invert_parts(cls, name_parts: List[str]) -> str:
        first = name_parts[0]
        last = name_parts[1]
        postnominals = cls.find_and_merge_postnominals(name_parts)
        return f'{last}, {first} {postnominals}'.strip()

    @staticmethod
    def find_and_merge_postnominals(name_parts: List[str]) -> str:
        postnominals = ''
        for np in name_parts[2:]:
            postnominals += np + ' '
        return postnominals


class RegularExpressions:
    @staticmethod
    def any_amount_of_whitespaces() -> str:
        return '\\s+'


class Honorifics:
    known_honorifics = ['Mr.', 'Mrs.']

    @classmethod
    def is_honorific(cls, s) -> bool:
        return s in cls.known_honorifics

File: name-inverter/name-inverter-java/src/main/java/pl/frognar/nameinverter/_02/NameInverter.java
package pl.frognar.nameinverter._02;

import com.google.common.collect.Lists;

import java.util.List;

public class NameInverter {
    public static String invert(String name) {
        List<String> nameParts = breakIntoPartsIgnoringWhitespace(name);
        if (nameParts.size() < 2)
            return nameParts.get(0);
        return invert(withoutHonorifics(nameParts));
    }

    private static List<String> breakIntoPartsIgnoringWhitespace(String name) {
        return Lists.newArrayList(name.trim().split(RegularExpressions.anyAmountOfWhitespace()));
    }

    private static List<String> withoutHonorifics(List<String> nameParts) {
        if (Honorifics.isHonorific(nameParts.get(0)))
            nameParts.remove(0);
        return nameParts;
    }

    private static String invert(List<String> nameParts) {
        String first = nameParts.get(0);
        String last = nameParts.get(1);
        String postnominals = findAndMergePostnominals(nameParts);
        return String.format("%s, %s %s", last, first, postnominals).trim();
    }

    private static String findAndMergePostnominals(List<String> nameParts) {
        StringBuilder postnominals = new StringBuilder();
        for (String np : nameParts.subList(2, nameParts.size()))
            postnominals.append(np).append(" ");
        return postnominals.toString();
    }
}

class RegularExpressions {
    public static String anyAmountOfWhitespace() {
        return "\\s+";
    }
}

class Honorifics {
    private static final List<String> knownHonorifics = List.of("Mr.");

    public static boolean isHonorific(String s) {
        return knownHonorifics.contains(s);
    }
}